Tools/WinMLRunner/src/EventTraceHelper.cpp

```diff
@@ -0,0 +1,559 @@
+    LPWSTR pwsEventGuid = NULL;
+    PBYTE pUserData = NULL;
+    PBYTE pEndOfUserData = NULL;
+    ULONGLONG TimeStamp = 0;
+    ULONGLONG Nanoseconds = 0;
+
+    // Skips the event if it is the event trace header. Log files contain this event
+    // but real-time sessions do not. The event contains the same information as
+    // the EVENT_TRACE_LOGFILE.LogfileHeader member that you can access when you open
+    // the trace.
+
+    if (IsEqualGUID(pEventRecord->EventHeader.ProviderId, EventTraceGuid) &&
+        pEventRecord->EventHeader.EventDescriptor.Opcode == EVENT_TRACE_TYPE_INFO)
+    {
+        // Skip this event.
+    }
+    else
+    {
+        // Process the event. The pEventRecord->UserData member is a pointer to
+        // the event specific data, if it exists.
+
+        status = GetEventInformation(pEventRecord, pInfo);
+
+        if (ERROR_SUCCESS != status)
+        {
+            printf("GetEventInformation failed with %lu\n", status);
+            goto cleanup;
+        }
+
+        if (DecodingSourceTlg == pInfo->DecodingSource)
+        {
+            FormatDataForCPUFallback(pEventRecord, pInfo);
+        }
+        else // Not handling any events other than Tlg type
+        {
+            // Do nothing
+        }
+    }
+
+cleanup:
+
+    if (pInfo)
+    {
+        free(pInfo);
+    }
+}
+
+ULONG WINAPI EventTraceHelper::BufferCallback(PEVENT_TRACE_LOGFILEW pLogFile) 
+{ 
+    return TRUE; 
+}
+
+void EventTraceHelper::ProcessEventTrace(PTRACEHANDLE traceHandle)
+{
+    try
+    {
+        auto status = ProcessTrace(
+            traceHandle, 1, 0,
+            0); // this call blocks until either the session is stopped or an exception is occurred in event_callback
+        if (status != ERROR_SUCCESS && status != ERROR_CANCELLED)
+        {
+            printf("ProcessTrace failed with %lu\n", status);
+        }
+    }
+    catch (exception ex)
+    {
+        printf("Exception when processing event trace: %s\n", ex.what());
+        // catch exceptions occurred in event_callback
+    }
+}
+
+void EventTraceHelper::Start()
+{
+    int bufferSize = sizeof(EVENT_TRACE_PROPERTIES) + (sizeof(LOGSESSION_NAME) + 1) * sizeof(wchar_t);
+    sessionProperties = static_cast<PEVENT_TRACE_PROPERTIES>(malloc(bufferSize));
+    ZeroMemory(sessionProperties, bufferSize);
+
+    GUID guid;
+    UuidCreate(&guid);
+    sessionProperties->Wnode.BufferSize = static_cast<ULONG>(bufferSize);
+    sessionProperties->Wnode.Guid = guid;
+    sessionProperties->Wnode.ClientContext = 0;
+    sessionProperties->Wnode.Flags = WNODE_FLAG_TRACED_GUID;
+    sessionProperties->LogFileMode = EVENT_TRACE_REAL_TIME_MODE;
+    sessionProperties->LoggerNameOffset = sizeof(EVENT_TRACE_PROPERTIES);
+
+    auto hr = StartTrace(static_cast<PTRACEHANDLE>(&sessionHandle), LOGSESSION_NAME, sessionProperties);
+    if (hr != ERROR_SUCCESS)
+    {
+        printf("Error starting event trace: Trace already started %s\n", GetLastError());

```

Comment:
Is this an error or a warning?

If it is an error, can we return; immediately?
Otherwise, can we say its a warning?

Reply:
Yeah a warning sounds more appropriate